### PR TITLE
Run empty reduction node level on data nodes

### PR DESCRIPTION
**Pull request by dnhatn**
This PR executes an empty reduction plan on data nodes at the node-level. With this change, we have the capability to replace this empty plan with an actual reduction plan later.

Relates #99498

### CODE REVIEW

**Comment by dnhatn** [comment]
run elasticsearch-ci/part-4

**Review by costin — COMMENTED** [review]
LGTM

**Comment by dnhatn** [comment]
test this please

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java [hunk 1]
```diff
@@ -665,17 +663,94 @@ private void onFailure(Exception e) {
...
+                internalSink,
+                request.configuration().pragmas().maxConcurrentShardsPerNode(),
+                collectedProfiles,
+                ActionListener.runBefore(cancelOnFailure(task, cancelled, refs.acquire()), responseHeadersCollector::collect)
+            );
+            dataNodeRequestExecutor.start();
+            // run the node-level reduction
+            var externalSink = exchangeService.getSinkHandler(externalId);
+            var exchangeSource = new ExchangeSourceHandler(1, esqlExecutor);
+            exchangeSource.addRemoteSink(internalSink::fetchPageAsync, 1);
```
Comment (astefan) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java:694]: I think it's confusing that we call the same method `addRemoteSink` for any kind of sink (remote/local). I know that we do not differentiate between them when it comes to functionality, but following the code is much harder like this. Maybe RemoteSink is a misnomer.

The ExchangeService is the one dealing with Sinks and Sources but it also has some static methods that receive a TransportService (implying the "remote" functionality). But, `ExchangeSinkHandler` is the one that creates a `LocalExchangeSink`, while the `ExchangeSourceHandler` can add a RemoteSink and create a `LocalExchangeSource`.
What I'm trying to point out is that it is difficult to establish what is "remote" and what is "local", who is in charge of managing these two types of sides.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java [hunk 2]
```diff
@@ -665,17 +663,94 @@ private void onFailure(Exception e) {
...
+            ActionListener<Void> reductionListener = cancelOnFailure(task, cancelled, refs.acquire());
+            runCompute(
```
[the remark this answers is not in the document]
Comment (astefan) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java:696]: Scratch that. The `reducePlan` is sink/source only. I misread.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java (hunk 1 above)
Comment (dnhatn) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java:694]: The RemoteSink doesn't necessarily have to be network . I opened https://github.com/elastic/elasticsearch/pull/106245 to rename `LocalExchangeSink` and `LocalExchangeSink`. These should not have the term `Local`. I will introduce an actual LocalExchange to reduce the number of buffers, and this should be no longer a concern.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dnhatn) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java:696]: The reduce plan is currently empty, containing only an exchange source and an exchange sink. We will insert the actual reduce plan between these.

```
\\_ExchangeSourceOperator
\\_ExchangeSinkOperator
```

**Review by nik9000 — APPROVED** [review]
Looks good to me once @astefan's good with it.

**Review by astefan — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
@costin @astefan @nik9000 Thanks for reviewing.

**Pull request merged**